[pyos][bug] pycodestyle false positive with E305, incorrect line numbers?

Part of: https://github.com/pyOpenSci/software-submission/issues/257#issuecomment-4523576886

I can't really tell what's going on here. 

So for this example from the quarto gallery: https://github.com/jjallaire/penguins-dashboard

I am getting this from pycodestyle:

```
penguins-dashboard/penguins.qmd:35:1: E302 expected 2 blank lines, found 1
penguins-dashboard/penguins.qmd:43:1: E305 expected 2 blank lines after class or function definition, found 0
...
penguins-dashboard/penguins.qmd:53:1: E302 expected 2 blank lines, found 0
```

When i look at the generated file, around that section I see this...

```python
@reactive.Calc
def filtered_penguins():
    data = penguins[penguins["species"].isin(input.species())]
    data = data[data["island"].isin(input.islands())]
    return data
# -
# -
# %% [python]
ui.input_select("dist", "Distribution:", choices=["kde", "hist"])  # noqa: E305,E501
ui.input_checkbox("rug", "Show rug marks", value = False)
# -
```

The strange part is that that first line with `@reactive.Calc` is line 35, and the function definition ends on line 39. The `ui.input_select` is on line 43, which is notably not a function definition. So It's a false positive of some kind? and on the wrong line maybe?

The `E305` error is raised for line 43 as it is the next code line subsequent to the function, and so it is says it expects 2 blank lines before it (i.e., between the end of the function and it).

This message should be being suppressed by the addition of `# noqa: E305` on line 43, but it is not.

<img width="773" height="234" alt="Image" src="https://github.com/user-attachments/assets/b596ad06-4041-436a-8d9d-20d10de81ee7" />

Flake8 also uses `E305` - it raises the valid issue for line 35, but does not have the same invalid issues as `pycodestyle`. When run on the penguins file (with other linting issues resolved):

```
=============================================================
Running flake8...
=============================================================
/home/amy/Documents/general/lintquarto/penguins.qmd:35:1: E302 expected 2 blank lines, found 1
=============================================================
Running pycodestyle...
=============================================================
/home/amy/Documents/general/lintquarto/penguins.qmd:35:1: E302 expected 2 blank lines, found 1
/home/amy/Documents/general/lintquarto/penguins.qmd:43:1: E305 expected 2 blank lines after class or function definition, found 0
/home/amy/Documents/general/lintquarto/penguins.qmd:53:1: E302 expected 2 blank lines, found 0
/home/amy/Documents/general/lintquarto/penguins.qmd:63:1: E302 expected 2 blank lines, found 0
```

--- a/docs/pages/tools/pycodestyle.qmd
+++ b/docs/pages/tools/pycodestyle.qmd
@@ -70,3 +70,29 @@ result = subprocess.run(
 print(result.stdout)
 ```
 :::
+
+## Known limitation for pycodestyle with lintquarto
+
+`lintquarto` works by converting `.qmd` files into temporary `.py` files. During this process, non-code regions (e.g., markdown and cell boundaries) are replaced with placeholder lines (`# -`) so that line numbers remain aligned between the original `.qmd` file and the generated `.py` file. Linters are then run on the generated Python file.
+
+This approach causes issues with `pycodestyle` for the E3xx "blank line/layout" rules (e.g., E302, E305). The inserted placeholder lines can cause `pycodestyle` to report false positive violations of these rules - as can be seen in the example above. While these false positives can be suppressed by `lintquarto` for other packages using `# noqa`,`pycodestyle` only supports `# noqa` for a limited subset of error codes and does not allow suppression of E3xx rules in this way. As a result, these false positives cannot be reliably ignored when using `pycodestyle` directly. This is a known limitation ([#142](https://github.com/lintquarto/lintquarto/issues/142)).
+
+### Work around
+
+If you want to apply `pycodestyle` checks to Quarto files without these false positives, we recommend using a wrapper tool like **flake8** or **ruff**. These tools include `pycodestyle` rules but allow suppression of E3xx rules via `# noqa`, making it possible to ignore the conversion-related violations.
+
+[**Flake8:**](flake8.qmd) You can restrict checks to `pycodestyle` rules via `setup.cfg`:
+
+```
+# setup.cfg
+[flake8]
+select = E,W
+```
+
+[**Ruff:**](ruff.qmd) You can restrict checks to `pycodestyle` rules via `pyproject.toml`:
+
+```
+# pyproject.toml
+[tool.ruff.lint]
+extend-select = ["E", "W"]
+```

--- a/src/lintquarto/converter.py
+++ b/src/lintquarto/converter.py
@@ -211,15 +211,15 @@ class QmdToPyConverter:
             self.current_chunk_is_valuebox = False
             self.current_chunk_eval = None  # Reset after chunk ends
             if self.preserve_line_count:
-                self.py_lines.append("# -")
+                self._append_placeholder()
 
         # Check if it is within a python code chunk
         elif self.in_python:
             self._handle_python_chunk(line)
 
-        # For all other lines, set to # -
+        # For all other lines, set to placeholder
         elif self.preserve_line_count:
-            self.py_lines.append("# -")
+            self._append_placeholder()
 
     def _handle_python_chunk(self, line: str) -> None:
         """
@@ -360,7 +360,7 @@ class QmdToPyConverter:
         self.in_chunk_options = False
 
     def _append_placeholder(self) -> None:
-        """Append placeholder (`# -`) if preserving line count."""
+        """Append placeholder if preserving line count."""
         if self.preserve_line_count:
             self.py_lines.append("# -")
 
@@ -630,7 +630,6 @@ def convert_qmd_to_py(
         with qmd_path.open(encoding="utf-8") as f:
             qmd_lines = f.readlines()
 
-        # Iterate over lines, keeping python code, and setting rest to "# -"
         py_lines = converter.convert(qmd_lines=qmd_lines)
 
         # Write the output file

